File: typescript/packages/var-core/src/diagnostics.ts
import type { Span } from './span.ts'

export type Severity = 'error' | 'warning'

export type Diagnostic = {
  readonly severity: Severity
  readonly code: DiagnosticCode
  readonly message: string
  readonly span: Span
}

export type DiagnosticCode = 'ambiguous-match' | 'error-fence-without-step'

export type Candidate = {
  readonly expression: string
  readonly sourceFile: string
  readonly sourceLine: number
}

export type AmbiguousInput = {
  readonly text: string
  readonly span: Span
  readonly candidates: ReadonlyArray<Candidate>
}

export function ambiguousMatch(input: AmbiguousInput): Diagnostic {
  const lines = input.candidates
    .map((c) => `  '${c.expression}'    at ${c.sourceFile}:${c.sourceLine}`)
    .join('\n')
  return {
    severity: 'error',
    code: 'ambiguous-match',
    message: `Ambiguous step: "${input.text}"\nMatched by:\n${lines}`,
    span: input.span,
  }
}

// An `error` fence declares its example expected-to-fail, but the example has
// no runnable step to produce that failure (nothing matched, or the match was
// ambiguous). `span` points at the orphaned fence.
export function errorFenceWithoutStep(input: { readonly span: Span }): Diagnostic {
  return {
    severity: 'error',
    code: 'error-fence-without-step',
    message:
      'This `error` fence marks the example as expected-to-fail, but the example has no step to run.',
    span: input.span,
  }
}

File: ruby/packages/var-core/lib/oselvar/var/core/diagnostics.rb
# frozen_string_literal: true

module Oselvar
  module Var
    module Core
      # A planning/run diagnostic on the shared rail. code is one of
      # "ambiguous-match", "error-fence-without-step", "drift". Port of
      # diagnostics.ts.
      Diagnostic = Data.define(:code, :severity, :message, :span)
      Candidate = Data.define(:expression, :source_file, :source_line)
      AmbiguousInput = Data.define(:text, :span, :candidates)

      module Diagnostics
        module_function

        def ambiguous_match(input)
          lines = input.candidates.map do |c|
            "  '#{c.expression}'    at #{c.source_file}:#{c.source_line}"
          end.join("\n")
          Diagnostic.new(
            severity: 'error',
            code: 'ambiguous-match',
            message: "Ambiguous step: \"#{input.text}\"\nMatched by:\n#{lines}",
            span: input.span
          )
        end

        def drift_detected(name, span)
          Diagnostic.new(
            severity: 'error',
            code: 'drift',
            message: "This paragraph was an example and no longer matches any step (drift): \"#{name}\".\n" \
                     'Fix the step so it matches again, or accept it as prose (run in update mode).',
            span: span
          )
        end

        def error_fence_without_step(span)
          Diagnostic.new(
            severity: 'error',
            code: 'error-fence-without-step',
            message: 'This `error` fence marks the example as expected-to-fail, ' \
                     'but the example has no step to run.',
            span: span
          )
        end
      end
    end
  end
end
